DIPs/DIP1045.md

```diff
@@ -0,0 +1,55 @@
+# Disallow Unsafe Declarations
+
+| Field           | Value                                                           |
+|-----------------|-----------------------------------------------------------------|
+| DIP:            |                                                                 |
+| Review Count:   |                                                                 |
+| Author:         | [Mark](markleaf131313@gmail.com)                                |
+| Implementation: |                                                                 |
+| Status:         | Draft                                                           |
+
+## Abstract
+
+Disallow extern C and C++ function declarations from being marked `@safe` or `@trusted`.
+
+## Contents
+* [Rationale](#rationale)
+* [Prior Work](#prior-work)
+* [Description](#description)
+* [Breaking Changes and Deprecations](#breaking-changes-and-deprecations)
+* [Reference](#reference)
+* [Copyright & License](#copyright--license)
+* [Reviews](#reviews)
+
+## Rationale
+
+Foreign code that is interfaced cannot be guaranteed to be safe. It should always be assumed as unsafe. Especially C declarations as the types used are not mangled into the name of the function. It would be possible to link to a C function with the incorrect parameter types.
```

Comment:
> It should always be assumed as unsafe.

That's the thesis of the DIP, the rationale section should explain in detail why that assumption is useful.

> It would be possible to link to a C function with the incorrect parameter types.

You can also link a D function with different signature by giving a false mangle.

Once the OS / build system / foreign code is compromised, there's nothing the programming language can do, so I would say it's best to assume those behave well, because if they don't, all bets are off. 

See also: https://forum.dlang.org/post/dcwzlotridqvrynzpxag@forum.dlang.org

Reply:
> Once the OS / build system / foreign code is compromised, there's nothing the programming language can do, so I would say _it's best to assume those behave well, because if they don't, all bets are off._

I’d use the last part to come to the very opposite conclusion. `@safe` means that the compiler checked it up to `@trusted` calls. Ideally, this means functions with `pragma(mangle,…)` should not be allowed to be marked `@safe` because it simply may not be. If your implementation is `@safe`, make a shallow wrapper like this:
```D
pragma(mangle, "body")
extern(C) void body_func() @trusted => safe_body();
void safe_body() @safe { … }
```
It is a little lengthy, but it’s the shortest way which is honest.

Reply:
I tend to think a function definition should always be allowed to be `@safe`. Only a declaration without a body should be limited to `@trusted` or `@system`, if it has non-D ABI or custom mangling. Binary mangling of `@trusted` and `@safe` should be merged so they would link to each other even in D mangling.

It still does mean that a false `@safe` D mangling could be given at the definition site without using `@trusted`, but that's a small enough hole that i think it's better than requiring the `safe_body` workaround. If we require the workaround, people are likely to just implement `body_func` with a `@trusted` body.